input/listen.go

```diff
@@ -3,45 +3,72 @@ package input
 import (
 	"bytes"
 	"net"
+	"time"
 )
 
 func listen(addr string, handler Handler) error {
-	laddr, err := net.ResolveTCPAddr("tcp", addr)
+	// listeners are set up outside of accept* here so they can interrupt startup
+	listener, err := listenTcp(addr)
 	if err != nil {
 		return err
 	}
-	l, err := net.ListenTCP("tcp", laddr)
-	if err != nil {
-		return err
-	}
-	log.Notice("listening on %v/tcp", laddr)
-	go acceptTcp(l, handler)
 
-	udp_addr, err := net.ResolveUDPAddr("udp", addr)
+	udp_conn, err := listenUdp(addr)
 	if err != nil {
 		return err
 	}
-	udp_conn, err := net.ListenUDP("udp", udp_addr)
-	if err != nil {
-		return err
-	}
-	log.Notice("listening on %v/udp", udp_addr)
-	go acceptUdp(udp_conn, handler)
+
+	go acceptTcp(addr, listener, handler)
+	go acceptUdp(addr, udp_conn, handler)
 
 	return nil
 }
 
-func acceptTcp(l *net.TCPListener, handler Handler) {
+func listenTcp(addr string) (*net.TCPListener, error) {
+	laddr, err := net.ResolveTCPAddr("tcp", addr)
+	if err != nil {
+		return nil, err
+	}
+	return net.ListenTCP("tcp", laddr)
+}
+
+func acceptTcp(addr string, listener *net.TCPListener, handler Handler) {
+	var err error
+	l := listener
 	for {
-		// wait for a tcp connection
-		c, err := l.AcceptTCP()
-		if err != nil {
+		log.Notice("listening on %v/tcp", addr)
+
+		for {
+			// wait for a tcp connection
+			c, err := l.AcceptTCP()
+			if err != nil {
+				log.Error(err.Error())
+				break
+			}
+			log.Debug("listen.go: tcp connection from %v", c.RemoteAddr())
+			// handle the connection
+			go acceptTcpConn(c, handler)
+		}
+
+		log.Notice("error accepting on %v/tcp, closing connection", addr)
```

Comment:
personally i'm not a fan of first logging the error without context, then a second log entry with only the context and the action we're taking
why not just 1 log.Error call with the context, the error, and tthe action we're taking?